homeassistant/components/vicare/water_heater.py

```diff
@@ -0,0 +1,168 @@
+    VICARE_MODE_DHWANDHEATING: OPERATION_MODE_ON,
+    VICARE_MODE_FORCEDREDUCED: OPERATION_MODE_OFF,
+    VICARE_MODE_FORCEDNORMAL: OPERATION_MODE_ON,
+    VICARE_MODE_OFF: OPERATION_MODE_OFF,
+}
+
+HA_TO_VICARE_HVAC_DHW = {
+    OPERATION_MODE_OFF: VICARE_MODE_OFF,
+    OPERATION_MODE_ON: VICARE_MODE_DHW,
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_CIRCUIT): int,
+        vol.Optional(CONF_NAME, default="ViCare"): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Create the ViCare water_heater devices."""
+    if config.get(CONF_CIRCUIT) is None:
+        vicare_api = Device(
+            config[CONF_USERNAME], config[CONF_PASSWORD], "/tmp/vicare_token.save"
+        )
+    else:
+        vicare_api = Device(
+            config[CONF_USERNAME],
+            config[CONF_PASSWORD],
+            "/tmp/vicare_token.save",
+            config[CONF_CIRCUIT],
+        )
+    add_entities([ViCareWater(f"{config[CONF_NAME]} Water", vicare_api)])
+
+
+class ViCareWater(WaterHeaterDevice):
+    """Representation of the ViCare domestic hot water device."""
+
+    def __init__(self, name, api):
+        """Initialize the DHW water_heater device."""
+        self._name = name
+        self._state = None
+        self._api = api
+        self._target_temperature = None
+        self._current_temperature = None
+        self._current_mode = None
+
+    def update(self):
+        """Let HA know there has been an update from the ViCare API."""
+        current_temperature = self._api.getDomesticHotWaterStorageTemperature()
+        if current_temperature is not None and current_temperature != "error":
+            self._current_temperature = current_temperature
+        else:
+            self._current_temperature = -1
+
+        self._target_temperature = self._api.getDomesticHotWaterConfiguredTemperature()
+
+        self._current_mode = self._api.getActiveMode()
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return SUPPORT_FLAGS_HEATER
+
+    @property
+    def name(self):
+        """Return the name of the water_heater device."""
+        return self._name
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        return self._current_temperature
+
+    @property
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        return self._target_temperature
+
+    def set_temperature(self, **kwargs):
+        """Set new target temperatures."""
+        temp = kwargs.get(ATTR_TEMPERATURE)
+        if temp is not None:
+            self._target_temperature = temp
```

Comment:
`update` will be called at the end of the service call, so we shouldn't update any state holding instance attributes in the actuator methods.